File: backend/app/schemas/policy.py
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class PolicyCard(BaseModel):
    plcy_no: str
    plcy_nm: str
    category: Optional[str] = None
    keywords: List[str] = []  # 통제어휘(목록 카드에도 제공)
    region: str
    org: Optional[str] = None
    summary: Optional[str] = None
    benefit: Optional[str] = None
    age_label: str  # 연령 표시(범위/한쪽/무관/확인필요 폴백)
    dday: str
    days: Optional[int] = None
    views: int
    is_always_open: bool
    sprt_arvl_seq_yn: Optional[bool] = None  # 선착순 여부(뱃지)
    apply_end_date: Optional[date] = None
    aply_url_addr: str  # 신청 URL (없으면 온통청년 상세 페이지로 폴백)


class PolicyListResponse(BaseModel):
    total: int
    page: int
    size: int
    items: List[PolicyCard]


class Eligibility(BaseModel):
    # 지원대상 구조화 (각 항목은 의미있을 때만 채워짐 — '제한없음'은 생략)
    age: str                              # 연령 (age_label)
    region: str                           # 거주지역
    income: Optional[str] = None          # 소득 조건
    marriage: Optional[str] = None        # 혼인 (제한없음 생략)
    job: Optional[str] = None             # 취업상태 요건
    education: Optional[str] = None        # 학력 요건
    major: Optional[str] = None            # 전공 요건
    specialization: Optional[str] = None   # 특화분야
    additional: Optional[str] = None       # 추가 자격조건 원문


class PolicyDetail(BaseModel):
    plcy_no: str
    plcy_nm: str
    category: Optional[str] = None
    lclsf_nm: Optional[str] = None
    mclsf_nm: Optional[str] = None
    plcy_expln_cn: Optional[str] = None
    plcy_sprt_cn: Optional[str] = None

    region: str
    region_sido: List[str]
    is_nationwide: bool
    keywords: List[str]

    sprt_trgt_min_age: Optional[int] = None
    sprt_trgt_max_age: Optional[int] = None
    age_label: str  # 연령 표시 폴백(범위/한쪽/무관/확인필요)
    earn_cnd_se_cd: Optional[str] = None
    earn_cnd_se_nm: Optional[str] = None  # 코드 라벨(code 테이블)
    earn_min_amt: Optional[int] = None
    earn_max_amt: Optional[int] = None
    mrg_stts_cd: Optional[str] = None
    mrg_stts_nm: Optional[str] = None  # 코드 라벨(code 테이블)

    aply_prd_se_cd: Optional[str] = None
    aply_prd_se_nm: Optional[str] = None  # 코드 라벨(code 테이블)
    is_always_open: bool
    apply_start_date: Optional[date] = None
    apply_end_date: Optional[date] = None
    dday: str
    days: Optional[int] = None
    is_active: bool  # 온통청년 현재 목록 여부(FALSE=종료/미제공 — 만료 정책 구분용)
    sprt_arvl_seq_yn: Optional[bool] = None   # 선착순 여부
    sprt_scl_lmt_yn: Optional[bool] = None     # 지원 규모(인원) 제한 여부
    sprt_scl_cnt: Optional[int] = None         # 지원 규모/인원 (0/None=미지정)
    plcy_pvsn_mthd_cd: Optional[str] = None    # 지원 방식 코드
    plcy_pvsn_mthd_nm: Optional[str] = None    # 지원 방식 라벨(프로그램/보조금/바우처 등)

    sprvsn_inst_cd_nm: Optional[str] = None
    aply_url_addr: Optional[str] = None
    views: int
    frst_reg_dt: Optional[datetime] = None
    last_mdfcn_dt: Optional[datetime] = None

    eligibility: Eligibility           # 지원대상 구조화
    target: str                        # 지원대상 조합 문자열(프론트 즉시 호환)

    # 상세 안내 섹션 (raw_data 추출 + placeholder 정제 → 빈값은 None)
    apply_method: Optional[str] = None    # 신청 방법
    documents: Optional[str] = None        # 제출 서류
    screening: Optional[str] = None        # 심사/선정 방법
    etc_notes: Optional[str] = None        # 기타 사항
    oper_inst_nm: Optional[str] = None     # 운영기관명(주관기관과 별개)
    contact: Optional[str] = None          # 문의처/담당 부서
    ref_urls: List[str] = []               # 참고 링크
    biz_period: Optional[str] = None       # 사업 기간(신청기간과 별개)


class ShareCreateRequest(BaseModel):
    plcy_nos: List[str] = Field(min_length=1, max_length=50)  # 공유할 찜 목록(클라이언트 스냅샷)


class ShareCreateResponse(BaseModel):
    share_code: str
    expires_at: datetime

File: frontend/lib/api/policy.ts
import { useQuery, useInfiniteQuery, keepPreviousData, type QueryClient } from "@tanstack/react-query";
import { mockDetail, MOCK_CARDS } from "./mock";
import { useAuthStore } from "@/lib/store/authStore";

const USE_MOCK = process.env.NEXT_PUBLIC_USE_MOCK === "true";

// 로그인 상태면 토큰을 실어 보내 sort=recommended 등 개인화 정렬이 동작하게 함(비로그인도 정상 동작)
const fetcher = (url: string, signal?: AbortSignal) => {
  const token = useAuthStore.getState().accessToken;
  return fetch(url, {
    signal,
    ...(token && { headers: { Authorization: `Bearer ${token}` } }),
  }).then((r) => {
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    return r.json();
  });
};

export interface PolicyDetailData {
  plcy_no: string;
  plcy_nm: string;
  category?: string | null;
  lclsf_nm?: string | null;
  mclsf_nm?: string | null;
  plcy_expln_cn?: string | null;
  plcy_sprt_cn?: string | null;
  region: string;
  region_sido: string[];
  is_nationwide: boolean;
  keywords: string[];
  sprt_trgt_min_age?: number | null;
  sprt_trgt_max_age?: number | null;
  earn_cnd_se_cd?: string | null;
  earn_min_amt?: number | null;
  earn_max_amt?: number | null;
  mrg_stts_cd?: string | null;
  aply_prd_se_cd?: string | null;
  is_always_open: boolean;
  apply_start_date?: string | null;
  apply_end_date?: string | null;
  dday: string;
  days?: number | null;
  sprvsn_inst_cd_nm?: string | null;
  aply_url_addr?: string | null;
  views: number;
  frst_reg_dt?: string | null;
  last_mdfcn_dt?: string | null;
  age_label?: string | null;
  target?: string | null;
  apply_method?: string | null;
  documents?: string | null;
  screening?: string | null;
  etc_notes?: string | null;
  oper_inst_nm?: string | null;
  contact?: string | null;
  ref_urls?: string[] | null;
  biz_period?: string | null;
}

export interface PolicyCardData {
  plcy_no: string;
  plcy_nm: string;
  category?: string | null;
  region: string;
  org?: string | null;
  summary?: string | null;
  benefit?: string | null;
  dday: string;
  days?: number | null;
  views: number;
  is_always_open: boolean;
  sprt_arvl_seq_yn?: boolean | null;
  apply_end_date?: string | null;
  keywords?: string[] | null;
  aply_url_addr?: string | null;
  age_label?: string | null;
}

export interface PolicyListResponse {
  total: number;
  page: number;
  size: number;
  items: PolicyCardData[];
}

// 마감(dday === "마감")인 정책은 가장 아래로, 나머지는 D-day 오름차순(남은 일수 적은 순)으로 정렬
// days가 없는 상시모집 정책은 남은 일수가 무한한 것으로 취급해 마감되지 않은 정책 중 가장 뒤에 위치
export function sortPoliciesByDeadline<T extends { dday: string; days?: number | null }>(items: T[]): T[] {
  return [...items].sort((a, b) => {
    const aClosed = a.dday === "마감";
    const bClosed = b.dday === "마감";
    if (aClosed !== bClosed) return aClosed ? 1 : -1;
    if (aClosed) return 0;
    return (a.days ?? Infinity) - (b.days ?? Infinity);
  });
}

const BASE = process.env.NEXT_PUBLIC_API_URL;

export interface PolicyListParams {
  q?: string;
  category?: string[];
  keywords?: string[];
  sido?: string;
  age?: number;
  size?: number;
  // "recommended": 백엔드 sort 패턴(popular|deadline|recent)에 아직 없음 — 지원 전까지 사용 금지.
  // 복구 시 workflow/9_feat/phase6_설문정책매칭.md 참고.
  sort?: "recent" | "popular" | "deadline" | "recommended";
  applicable?: boolean;
  job?: string[];
}

function buildPolicyQuery(params: PolicyListParams & { page?: number }) {
  const query = new URLSearchParams();
  if (params.q) query.set("q", params.q);
  params.category?.forEach((c) => query.append("category", c));
  params.keywords?.forEach((k) => query.append("keywords", k));
  params.job?.forEach((j) => query.append("job", j));
  if (params.sido) query.set("sido", params.sido);
  if (params.age != null) query.set("age", String(params.age));
  if (params.page) query.set("page", String(params.page));
  if (params.size) query.set("size", String(params.size));
  if (params.sort) query.set("sort", params.sort);
  if (params.applicable) query.set("applicable", "true");
  return query;
}

function policyDetailQueryOptions(policyId: string) {
  return {
    queryKey: ["policy", "detail", policyId] as const,
    queryFn: ({ signal }: { signal?: AbortSignal }) =>
      fetcher(`${BASE}/policy/get/policy/${policyId}`, signal),
    staleTime: 60_000,
  };
}

// 카드에 hover하는 순간 상세 데이터를 미리 캐시에 채워, 클릭 후 진입 시 로딩 없이 바로 렌더되게 함
export function prefetchPolicyDetail(queryClient: QueryClient, policyId: string) {
  if (USE_MOCK || !policyId) return;
  return queryClient.prefetchQuery(policyDetailQueryOptions(policyId));
}

export function usePolicyDetail(policyId: string | null) {
  const { data, error, isLoading } = useQuery<PolicyDetailData>({
    ...policyDetailQueryOptions(policyId ?? ""),
    enabled: !USE_MOCK && !!policyId,
  });

  if (USE_MOCK) {
    return { policy: policyId ? mockDetail(policyId) : null, error: null, isLoading: false };
  }
  return { policy: data ?? null, error: error ?? null, isLoading };
}

export function usePolicyList(params: (PolicyListParams & { page?: number }) | null) {
  const url = `${BASE}/policy/get/search?${params ? buildPolicyQuery(params).toString() : ""}`;

  const { data, error, isLoading } = useQuery<PolicyListResponse>({
    queryKey: ["policy", "list", params],
    queryFn: ({ signal }) => fetcher(url, signal),
    enabled: params !== null && !USE_MOCK,
    staleTime: 5 * 60_000,
    gcTime: 30 * 60_000,
    placeholderData: keepPreviousData,
  });

  if (USE_MOCK) {
    if (params === null) return { data: null, error: null, isLoading: false };
    let items = MOCK_CARDS;
    if (params.q) {
      const q = params.q.toLowerCase();
      items = items.filter(
        (c) =>
          c.plcy_nm.toLowerCase().includes(q) ||
          c.summary?.toLowerCase().includes(q) ||
          c.keywords?.some((k) => k.toLowerCase().includes(q))
      );
    }
    return { data: { total: items.length, page: 1, size: params.size ?? 20, items }, error: null, isLoading: false };
  }
  return { data: data ?? null, error: error ?? null, isLoading };
}

// 스크롤 시 다음 페이지를 이어붙여 전체 정책을 가져오는 무한스크롤용 훅
export function useInfinitePolicyList(params: PolicyListParams | null) {
  const size = params?.size ?? 20;

  const query = useInfiniteQuery<PolicyListResponse>({
    queryKey: ["policy", "list", "infinite", params],
    queryFn: ({ pageParam, signal }) =>
      fetcher(`${BASE}/policy/get/search?${buildPolicyQuery({ ...params, size, page: pageParam as number }).toString()}`, signal),
    initialPageParam: 1,
    getNextPageParam: (lastPage) =>
      lastPage.page * lastPage.size < lastPage.total ? lastPage.page + 1 : undefined,
    enabled: params !== null && !USE_MOCK,
    staleTime: 5 * 60_000,
    gcTime: 30 * 60_000,
    placeholderData: keepPreviousData,
  });

  if (USE_MOCK) {
    let items = params === null ? [] : MOCK_CARDS;
    if (params?.q) {
      const q = params.q.toLowerCase();
      items = items.filter(
        (c) =>
          c.plcy_nm.toLowerCase().includes(q) ||
          c.summary?.toLowerCase().includes(q) ||
          c.keywords?.some((k) => k.toLowerCase().includes(q))
      );
    }
    return {
      items,
      total: items.length,
      error: null,
      isLoading: false,
      hasNextPage: false,
      isFetchingNextPage: false,
      fetchNextPage: () => {},
    };
  }

  return {
    items: query.data?.pages.flatMap((p) => p.items) ?? [],
    total: query.data?.pages[0]?.total ?? 0,
    error: query.error ?? null,
    isLoading: query.isLoading,
    hasNextPage: !!query.hasNextPage,
    isFetchingNextPage: query.isFetchingNextPage,
    fetchNextPage: query.fetchNextPage,
  };
}
